straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,316 @@ def find_n_competing(peaks, windows, fraction):
+
+        dtype = s1_time_shadow_dtype + s2_time_shadow_dtype + s2_position_shadow_dtype + nearest_dtype + strax.time_fields
+        return dtype
+
+    @property
+    def shadowdtype(self):
+        dtype = []
+        dtype += [('shadow', np.float32), ('pre_dt', np.int64)]
+        dtype += [('pre_x', np.float32), ('pre_y', np.float32)]
+        dtype += [('nearest_dt', np.int64)]
         return dtype
 
     def compute(self, peaks):
-        roi_shadow = np.zeros(len(peaks), dtype=strax.time_fields)
-        roi_shadow['time'] = peaks['center_time'] - self.time_window_backward
-        roi_shadow['endtime'] = peaks['center_time'] - self.n_drift_time
-
-        mask_pre_s2 = peaks['area'] > self.s2_threshold
-        mask_pre_s2 &= peaks['type'] == 2
-        split_peaks = strax.touching_windows(peaks[mask_pre_s2], roi_shadow)
-        res = np.zeros(len(peaks), self.dtype)
-        res['pre_s2_x'] = np.nan
-        res['pre_s2_y'] = np.nan
-        if len(peaks):
-            self.compute_shadow(peaks, peaks[mask_pre_s2], split_peaks, self.exponent, res)
-
-        res['time'] = peaks['time']
-        res['endtime'] = strax.endtime(peaks)
-        return res
+        return self.compute_shadow(peaks, peaks)
+
+    def compute_shadow(self, peaks, current_peak):
+        # 1. Define time window for each peak, we will find previous peaks within these time windows
+        roi_shadow = np.zeros(len(current_peak), dtype=strax.time_fields)
+        roi_shadow['time'] = current_peak['center_time'] - self.time_window_backward
+        roi_shadow['endtime'] = current_peak['center_time']
+
+        # 2. Calculate S2 position shadow, S2 time shadow, and S1 time shadow
+        result = np.zeros(len(current_peak), self.dtype)
+        for key, is_position in zip(['s2_position', 's2_time', 's1_time'], [True, False, False]):
+            type_str = key.split('_')[0]
+            stype = 2 if 's2' in key else 1
+            mask_pre = (peaks['type'] == stype) & (peaks['area'] > self.threshold[key])
+            split_peaks = strax.touching_windows(peaks[mask_pre], roi_shadow)
+            array = np.zeros(len(current_peak), np.dtype(self.shadowdtype))
+
+            # Initialization
+            array['pre_x'] = np.nan
+            array['pre_y'] = np.nan
+            array['pre_dt'] = self.time_window_backward
+            # Shadow is set to be the lowest possible value
+            if 'time' in key:
+                array['shadow'] = self.threshold[key] * array['pre_dt'] ** self.exponent
+            else:
+                array['shadow'] = 0
+            array['nearest_dt'] = self.time_window_backward
+
+            # Calculating shadow, the Major of the plugin. Only record the previous peak casting the largest shadow
+            if len(current_peak):
+                self.peaks_shadow(current_peak, peaks[mask_pre], split_peaks, self.exponent, array, is_position, self.getsigma(self.sigma, self.baseline, current_peak['area']))
```

Comment:
Can you use line breaks after each argument? I have already a very large screen and it still does not fit :D In case you are using pycharm you can set a max line number. If I recall correctly we settled on something like 130. Maybe also check the rest f your code you have some functions with very long arguments.

Reply:
I will check them. :)